File: modules/@angular/core/testing/fake_async.ts
import {BaseException} from '../index';
import {getTestInjector} from './test_injector';

let _FakeAsyncTestZoneSpecType = (Zone as any /** TODO #9100 */)['FakeAsyncTestZoneSpec'];

/**
 * Wraps a function to be executed in the fakeAsync zone:
 * - microtasks are manually executed by calling `flushMicrotasks()`,
 * - timers are synchronous, `tick()` simulates the asynchronous passage of time.
 *
 * If there are any pending timers at the end of the function, an exception will be thrown.
 *
 * Can be used to wrap inject() calls.
 *
 * ## Example
 *
 * {@example testing/ts/fake_async.ts region='basic'}
 *
 * @param fn
 * @returns {Function} The function wrapped to be executed in the fakeAsync zone
 */
export function fakeAsync(fn: Function): Function {
  if (Zone.current.get('FakeAsyncTestZoneSpec') != null) {
    throw new BaseException('fakeAsync() calls can not be nested');
  }

  let fakeAsyncTestZoneSpec = new _FakeAsyncTestZoneSpecType();
  let fakeAsyncZone = Zone.current.fork(fakeAsyncTestZoneSpec);

  return function(...args: any[] /** TODO #9100 */) {
    let res = fakeAsyncZone.run(() => {
      let res = fn(...args);
      flushMicrotasks();
      return res;
    });

    if (fakeAsyncTestZoneSpec.pendingPeriodicTimers.length > 0) {
      throw new BaseException(
          `${fakeAsyncTestZoneSpec.pendingPeriodicTimers.length} ` +
          `periodic timer(s) still in the queue.`);
    }

    if (fakeAsyncTestZoneSpec.pendingTimers.length > 0) {
      throw new BaseException(
          `${fakeAsyncTestZoneSpec.pendingTimers.length} timer(s) still in the queue.`);
    }
    return res;
  };
}

function _getFakeAsyncZoneSpec(): any {
  let zoneSpec = Zone.current.get('FakeAsyncTestZoneSpec');
  if (zoneSpec == null) {
    throw new Error('The code should be running in the fakeAsync zone to call this function');
  }
  return zoneSpec;
}

/**
 * Clear the queue of pending timers and microtasks.
 * Tests no longer need to call this explicitly.
 *
 * @deprecated
 */
export function clearPendingTimers(): void {
  // Do nothing.
}

/**
 * Simulates the asynchronous passage of time for the timers in the fakeAsync zone.
 *
 * The microtasks queue is drained at the very start of this function and after any timer callback
 * has been executed.
 *
 * ## Example
 *
 * {@example testing/ts/fake_async.ts region='basic'}
 *
 */
export function tick(millis: number = 0): void {
  _getFakeAsyncZoneSpec().tick(millis);
}

/**
 * Discard all remaining periodic tasks.
 */
export function discardPeriodicTasks(): void {
  let zoneSpec = _getFakeAsyncZoneSpec();
  let pendingTimers = zoneSpec.pendingPeriodicTimers;
  zoneSpec.pendingPeriodicTimers.length = 0;
}

/**
 * Flush any pending microtasks.
 */
export function flushMicrotasks(): void {
  _getFakeAsyncZoneSpec().flushMicrotasks();
}

File: modules/@angular/core/testing/fake_async.dart
library testing.fake_async;

import 'dart:async' show runZoned, ZoneSpecification;
import 'package:quiver/testing/async.dart' as quiver;
import 'package:angular2/src/facade/exceptions.dart' show BaseException;

const _u = const Object();

quiver.FakeAsync _fakeAsync = null;

/**
 * Wraps the [fn] to be executed in the fakeAsync zone:
 * - microtasks are manually executed by calling [flushMicrotasks],
 * - timers are synchronous, [tick] simulates the asynchronous passage of time.
 *
 * If there are any pending timers at the end of the function, an exception
 * will be thrown.
 *
 * Can be used to wrap inject() calls.
 *
 * Returns a `Function` that wraps [fn].
 */
Function fakeAsync(Function fn) {
  if (_fakeAsync != null) {
    throw 'fakeAsync() calls can not be nested';
  }

  return ([a0 = _u,
           a1 = _u,
           a2 = _u,
           a3 = _u,
           a4 = _u,
           a5 = _u,
           a6 = _u,
           a7 = _u,
           a8 = _u,
           a9 = _u]) {
    // runZoned() to install a custom exception handler that re-throws
    return runZoned(() {
      return new quiver.FakeAsync().run((quiver.FakeAsync async) {
        try {
          _fakeAsync = async;
          List args = [a0, a1, a2, a3, a4, a5, a6, a7, a8, a9]
              .takeWhile((a) => a != _u)
              .toList();
          var res = Function.apply(fn, args);
          _fakeAsync.flushMicrotasks();

          if (async.periodicTimerCount > 0) {
            throw new BaseException('${async.periodicTimerCount} periodic '
                'timer(s) still in the queue.');
          }

          if (async.nonPeriodicTimerCount > 0) {
            throw new BaseException('${async.nonPeriodicTimerCount} timer(s) '
                'still in the queue.');
          }

          return res;
        } finally {
          _fakeAsync = null;
        }
      });
    },
        zoneSpecification: new ZoneSpecification(
            handleUncaughtError: (self, parent, zone, error, stackTrace) =>
                throw error));
  };
}

/**
 * Simulates the asynchronous passage of [millis] milliseconds for the timers
 * in the fakeAsync zone.
 *
 * The microtasks queue is drained at the very start of this function and after
 * any timer callback has been executed.
 */
void tick([int millis = 0]) {
  _assertInFakeAsyncZone();
  var duration = new Duration(milliseconds: millis);
  _fakeAsync.elapse(duration);
}

/**
 * This is not needed in Dart. Because quiver correctly removes a timer when
 * it throws an exception.
 */
void clearPendingTimers() {}

/**
 * Flush any pending microtasks.
 */
void flushMicrotasks() {
  _assertInFakeAsyncZone();
  _fakeAsync.flushMicrotasks();
}

void _assertInFakeAsyncZone() {
  if (_fakeAsync == null) {
    throw new BaseException('The code should be running in the fakeAsync zone '
        'to call this function');
  }
}
